File: SecureBlackbox Samples/Archive Writer/php/archive_writer.php
<?php
/*
 * SecureBlackbox 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/secureblackbox_archivewriter.php');
require_once('../include/secureblackbox_certificatemanager.php');
require_once('../include/secureblackbox_const.php');
?>
<style>
table { width: 100% !important; }
td { white-space: nowrap; }
td input { width: 100%; }
td:last-child { width: 100%; }
</style>

<div width="90%">
  <form method=POST>
    <h2>Archive Writer Demo</h2>
    
    <b>General Options</b><br/><br/>
    <table>
      <tr><td>Archive File:</td><td><input type=text name=archiveFile value=""></td></tr>
      <tr><td>Archive password:</td><td><input type=password name=archivePass value=""></td></tr>
    </table>

    <br/><br/><b>Archive Options</b><br/><br/>
    <table>
      <tr>
        <td>Archive type:</td>
        <td>
          <select name="arcType">
            <option value="ARCHIVEWRITER_ARCHIVETYPE_ZIP">Zip</option>
            <option value="ARCHIVEWRITER_ARCHIVETYPE_GZIP">GZip</option>
            <option value="ARCHIVEWRITER_ARCHIVETYPE_BZIP_2">BZip2</option>
            <option value="ARCHIVEWRITER_ARCHIVETYPE_TAR">Tar</option>
            <option value="ARCHIVEWRITER_ARCHIVETYPE_TAR_GZIP">Tar_GZip</option>
            <option value="ARCHIVEWRITER_ARCHIVETYPE_TAR_BZIP_2">Tar_BZip2</option>
          </select>
        </td>
      </tr>
      <tr>
        <td>Compression level:</td>
        <td>
          <select name="compressLevel">
            <option value="1">1</option>
            <option value="2">2</option>
            <option value="3">3</option>
            <option value="4">4</option>
            <option value="5">5</option>
            <option selected value="6">6</option>
            <option value="7">7</option>
            <option value="8">8</option>
            <option value="9">9</option>
          </select>
        </td>
      </tr>
    </table>
    <br/>
    <br/>

    <b>Files</b>
    <p>Enter the path(s) for any file(s), one per line.</p>
    <br/>
    <textarea style="font-family: Arial, sans-serif; width: 100%" name=filePaths rows=10></textarea>
    <br/>
    <br/>

    <input type="submit" value="Pack" />
  </form>
</div><br/>

<?php

function translateArchiveType($arcType){
  switch($arcType){
    case "ARCHIVEWRITER_ARCHIVETYPE_ZIP":  return ARCHIVEWRITER_ARCHIVETYPE_ZIP;  break;
    case "ARCHIVEWRITER_ARCHIVETYPE_GZIP":  return ARCHIVEWRITER_ARCHIVETYPE_GZIP;  break;
    case "ARCHIVEWRITER_ARCHIVETYPE_BZIP_2":  return ARCHIVEWRITER_ARCHIVETYPE_BZIP_2;  break;
    case "ARCHIVEWRITER_ARCHIVETYPE_TAR":  return ARCHIVEWRITER_ARCHIVETYPE_TAR;  break;
    case "ARCHIVEWRITER_ARCHIVETYPE_TAR_GZIP":  return ARCHIVEWRITER_ARCHIVETYPE_TAR_GZIP;  break;
    case "ARCHIVEWRITER_ARCHIVETYPE_TAR_BZIP_2":  return ARCHIVEWRITER_ARCHIVETYPE_TAR_BZIP_2;  break;
    default:  return ARCHIVEWRITER_ARCHIVETYPE_UNKNOWN;  break;
  }
}

  if ($_SERVER['REQUEST_METHOD'] == "POST") {
    $archivewriter = new SecureBlackbox_ArchiveWriter();

    try {
      // General options
      $archiveFile = $_REQUEST['archiveFile'];
      $archivePass = $_REQUEST['archivePass'];

      $archivewriter->setCompressionLevel($_REQUEST['compressLevel']);

      if (strlen($archivePass) > 0)
      {
        $archivewriter->setEncryptionType(ARCHIVEWRITER_ENCRYPTIONTYPE_GENERIC);
        $archivewriter->setEncryptionPassword($archivePass);
      }

      $archivewriter->doCreateNew(translateArchiveType($_REQUEST['arcType']));

      // Add files
      $filePaths = trim($_REQUEST['filePaths']);
      if (strlen($filePaths) > 0) {
        $files = explode("\r\n", $filePaths);
        for($x = 0; $x < count($files); $x++){
          $file = $files[$x];
          if (is_file($file)) {
            $archivewriter->doAddFile(basename($file ), $file );
          }
          elseif (is_dir($file)) {
            $archivewriter->doAddFiles("", $file, TRUE);
          }
        }
      }

      // Save
      $archivewriter->doSave($archiveFile);
      echo "<h2>Archive successfully create</h2>";
    }
    catch (exception $e) {
      echo "<h2>Packing Failure (Details Below)</h2><p>" . $e->getMessage() . "</p>";
    }
  }
?>

File: SecureBlackbox Samples/Archive Writer/c++/archivewriter.cpp
/*
 * SecureBlackbox 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <io.h>
#include <iostream>

#include "../../include/secureblackbox.h"

namespace ArgParser {
    static char* optval(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc - 1; x++) {
            if (!strcmp(argv[x], option)) {
                return argv[x + 1];
            }
        }
        return (char*)"";
    }

    static bool optext(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc; x++) {
            if (!strcmp(argv[x], option)) {
                return true;
            }
        }
        return false;
    }
};

using namespace ArgParser;

void displayHelp(const char* errMes) {
    printf(
        "NAME\n"
        "  archivewriter -- SecureBlackbox ArchiveWriter Demo Application\n\n"
        "SYNOPSIS\n"
		"  archivewriter <-arctype archive_type> <-input input1 input2 ....> <-output output_file> [-pass encryption_password] [-enctype encryption_type]\n\n"
        "DESCRIPTION\n"
        "  ArchiveWriter demonstrates the usage of ArchiveWriter from SecureBlackbox.\n"
        "  Used to create archive.\n\n"
        "  The options are as follows:\n\n"
        "  -arctype      The type of archive (Required). Valid values:\n\n"
        "                  1 - AFT_ZIP\n"
        "                  2 - AFT_GZIP\n"
        "                  3 - AFT_BZIP_2\n"
        "                  4 - AFT_TAR\n"
        "                  5 - AFT_TAR_GZIP\n"
        "                  6 - AFT_TAR_BZIP_2\n\n"
        "  -input        List of files or directories compressing to archive (Required).\n\n"
        "  -output       Where the archive file will be saved (Required).\n\n"
        "  -pass         The password for the encryption.\n\n"
        "  -enctype      The encryption type. Valid values: \n\n"
        "                  0 - AET_DEFAULT\n"
        "                  1 - AET_NO_ENCRYPTION\n"
        "                  2 - AET_GENERIC\n"
        "                  3 - AET_WIN_ZIP\n"
        "                  4 - AET_STRONG\n\n"
        "EXAMPLES\n"
        "  archivewriter -arctype 1 -input C:\\archive\\helloworld.txt -output C:\\archive\\myarchive.zip \n\n"
        "  archivewriter -arctype 5 -input C:\\archive\\helloworld.txt C:\\archive\\temp -output C:\\archive\\myarchive.tar -enctype 2 -pass mypassword \n\n"
    );

	if (strlen(errMes) > 0)
	{
		printf("Error: %s \n\n", errMes);
	}
}

bool FileExists(const char* fname) {
    _finddata_t data;
    long nFind = _findfirst(fname, &data);
    if (nFind != -1) {
        _findclose(nFind);

        if (data.attrib & _A_SUBDIR) {
            return false;
        } else {
            return true;
        }
    }
    return false;
}

std::string getFileName(const char* s) {
    std::string str(s);
    const std::size_t found = str.find_last_of("/\\");
    if (found != std::string::npos) {
        str = str.substr(found + 1);
    }
    return str;
}

int main(int argc, char** argv) {
    ArchiveWriter archive;
    int arctype;

    // Validate input
    if (argc <= 1) {
        displayHelp("");
        goto done;
    }

    if (optext(argc, argv, "-arctype")) {
        arctype = atoi(optval(argc, argv, "-arctype"));
    } else {
        displayHelp("-arctype is required.");
        goto done;
    }

    if (!strcmp(optval(argc, argv, "-input"), "")) {
        displayHelp("-input is required.");
        goto done;
    }

    char* output = optval(argc, argv, "-output");
    if (!strcmp(output, "")) {
        displayHelp("-output is required.");
        goto done;
    }

    // Additional options
    if (optext(argc, argv, "-enctype")) {
        archive.SetEncryptionType(atoi(optval(argc, argv, "-enctype")));
        archive.SetEncryptionPassword(optval(argc, argv, "-pass"));
    }

    // Create archive
    if (archive.CreateNew(arctype)) goto err;

    // Add input files
    for (int x = 0; x < argc; x++)
        if (!strcmp(argv[x], "-input") && x != argc - 1) {
            for (int i = x + 1; i < argc; i++) {
                if (argv[i][0] == '-') break;

                if (FileExists(argv[i])) // file exists
                {
                    archive.AddFile(getFileName(argv[i]).c_str(), argv[i]);
                } else {
                    archive.AddFiles("", argv[i], true);
                }
            }

            break;
        }

    printf("List of files added to the archive\n");
    for (int x = 0; x < archive.GetFileCount(); x++) {
        printf("    %s\n", archive.GetFilePath(x));
    }
    printf("\n");

    // Save archive
    if (archive.Save(output)) goto err;

    printf("Archive created.\n\n");

err:
    if (archive.GetLastErrorCode()) {
        printf("Error: [%i] %s\n", archive.GetLastErrorCode(), archive.GetLastError());
    }
done:
	printf("Press Enter to exit the demo.\n");
    getchar();
}


